Use warns rather than raises to test for warnings

I get failures running the tests under pytest. These are from #17124 (@drybalka @Upabjojr) and are caused by using `raises` instead of `warns` to test for deprecation warnings.

The reason this is not seen under the standard test runner is that it turns warnings into exceptions. This is problematic because it means that the tests can't distinguish between raising an exception and emitting a warning. It also means that deprecated functions cannot be tested properly because the warning will always lead to an exception that aborts the function.
```
=============================================== FAILURES ================================================
__________________________________________ test_tensorsymmetry __________________________________________
[gw4] linux -- Python 3.6.3 /home/cosc/enojb/linux/current/sympy/36venv/bin/python

    def test_tensorsymmetry():
        with raises(SymPyDeprecationWarning):
>           tensorsymmetry([1]*2)
E           Failed: DID NOT RAISE <class 'sympy.utilities.exceptions.SymPyDeprecationWarning'>

sympy/tensor/tests/test_tensor.py:1970: Failed
____________________________________________ test_tensorhead ____________________________________________
[gw4] linux -- Python 3.6.3 /home/cosc/enojb/linux/current/sympy/36venv/bin/python

    def test_tensorhead():
        with raises(SymPyDeprecationWarning):
>           tensorhead('A', [])
E           Failed: DID NOT RAISE <class 'sympy.utilities.exceptions.SymPyDeprecationWarning'>

sympy/tensor/tests/test_tensor.py:1974: Failed
____________________________________________ test_TensorType ____________________________________________
[gw4] linux -- Python 3.6.3 /home/cosc/enojb/linux/current/sympy/36venv/bin/python

    def test_TensorType():
        with raises(SymPyDeprecationWarning):
            sym = TensorSymmetry.no_symmetry(1)
>           TensorType(sym, 2)

sympy/tensor/tests/test_tensor.py:1979: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sympy.tensor.tensor.TensorType'>, index_types = TensorSymmetry((), ((2),)), symmetry = 2
kw_args = {}

    def __new__(cls, index_types, symmetry, **kw_args):
        deprecate_TensorType()
>       assert symmetry.rank == len(index_types)
E       AttributeError: 'int' object has no attribute 'rank'

sympy/tensor/tensor.py:1487: AttributeError
```
The last failure here is possibly an example where the code that is not tested because of the deprecation exception may not work any more.

The correct way to do this is to use `warns_deprecated_sympy`:
```python
from sympy.utilities.pytest import warns_deprecated_sympy, ignore_warnings
from sympy.utilities.exceptions import SymPyDeprecationWarning

def test_deprecated_function():
    # Asserts that the warning is emitted.
    with warns_deprecated_sympy():
        deprecated_function(1)

    # Suppress the warning while still testing the function
    with ignore_warnings(SymPyDeprecationWarning):
        assert deprecated_function(1) == 1
        assert deprecated_function(2) == 2
```

I have already removed all examples of raises(Warning) from the codebase but they keep coming back which makes me wonder if there is some out of date documentation somewhere. Also it would probably be good to fix the test runner so that it no longer allows this.

@drybalka I see. I can update that test. Btw, how could I go about trying to fix 

```
cls = <class 'sympy.tensor.tensor.TensorType'>, index_types = TensorSymmetry((), ((2),)), symmetry = 2
 kw_args = {}
 
     def __new__(cls, index_types, symmetry, **kw_args):
         deprecate_TensorType()
      assert symmetry.rank == len(index_types)
 E       AttributeError: 'int' object has no attribute 'rank'
 
 sympy/tensor/tensor.py:1487: AttributeError
```

--- a/sympy/tensor/tests/test_tensor.py
+++ b/sympy/tensor/tests/test_tensor.py
@@ -1975,5 +1975,6 @@ def test_tensorhead():
 
 def test_TensorType():
     with warns_deprecated_sympy():
-        sym = TensorSymmetry.no_symmetry(1)
-        TensorType(sym, 2)
+        sym2 = TensorSymmetry.fully_symmetric(2)
+        Lorentz = TensorIndexType('Lorentz')
+        S2 = TensorType([Lorentz]*2, sym2)

--- a/sympy/tensor/tests/test_tensor.py
+++ b/sympy/tensor/tests/test_tensor.py
@@ -13,7 +13,7 @@ from sympy.tensor.tensor import TensorIndexType, tensor_indices, TensorSymmetry,
     riemann_cyclic_replace, riemann_cyclic, TensMul, tensor_heads, \
     TensorManager, TensExpr, TensorHead, canon_bp, \
     tensorhead, tensorsymmetry, TensorType
-from sympy.utilities.pytest import raises, XFAIL, ignore_warnings
+from sympy.utilities.pytest import raises, XFAIL, warns_deprecated_sympy, ignore_warnings
 from sympy.utilities.exceptions import SymPyDeprecationWarning
 from sympy.core.compatibility import range
 from sympy.matrices import diag
@@ -1966,14 +1966,14 @@ def test_rewrite_tensor_to_Indexed():
 
 
 def test_tensorsymmetry():
-    with raises(SymPyDeprecationWarning):
+    with warns_deprecated_sympy():
         tensorsymmetry([1]*2)
 
 def test_tensorhead():
-    with raises(SymPyDeprecationWarning):
+    with warns_deprecated_sympy():
         tensorhead('A', [])
 
 def test_TensorType():
-    with raises(SymPyDeprecationWarning):
+    with warns_deprecated_sympy():
         sym = TensorSymmetry.no_symmetry(1)
         TensorType(sym, 2)

